[OSDOCS#18859]: Added files for Degraded TNF (#110945)
openshift/openshift-docs: modules/recovering-a-failed-tnf-node.adoc

// Module included in the following assemblies:
// * installing_tnf/operating-degraded-two-node-fencing-cluster.adoc

:_mod-docs-content-type: PROCEDURE
[id="recovering-a-failed-tnf-node_{context}"]
= Recovering a failed TNF node

[role="_abstract"]
Recovery is automatic when the failed node is powered on and rejoins the network. 

[IMPORTANT]
====
Degraded operation is a temporary state. Restore the failed node as soon as possible to regain fault tolerance, fencing protection, and the ability to perform upgrades and maintenance. If the original node cannot be recovered, replace it.

A TNF cluster running on a single node indefinitely is at risk of:

* Complete cluster loss if the surviving node fails
* Inability to apply security updates or upgrades
* Certificate expiry if the cluster remains degraded beyond certificate validity periods
====

.Procedure

> . Power on the failed node and ensure it has network connectivity.

Reviewer: ```suggestion
 Power on the failed node, verify connectivity on all three network planes before the node can fully rejoin
   the cluster:

  - BMC/management network — The peer node must be able to reach this node's BMC address. Without BMC connectivity,
   fencing cannot protect the cluster if a subsequent failure occurs.
  - Cluster network — Bidirectional connectivity between nodes for Corosync membership (ports 5405-5407), Pacemaker
   management (port 2224), and etcd replication (ports 2379, 2380).
  - Kubernetes API — The recovering node must be able to reach the API server (port 6443).

```

// Module included in the following assemblies:
// * installing_tnf/operating-degraded-two-node-fencing-cluster.adoc

:_mod-docs-content-type: PROCEDURE
[id="recovering-a-failed-tnf-node_{context}"]
= Recovering a failed TNF node

[role="_abstract"]
Recovery is automatic when the failed node is powered on and rejoins the network. 
Degraded operation is a temporary state. Restore the failed node as soon as possible to regain fault tolerance, fencing protection, and the ability to perform upgrades and maintenance. If the original node cannot be recovered, replace it.

[IMPORTANT]
====

> 

Reviewer 2: ```suggestion
```

> A Two-Node with Fencing (TNF) cluster running on a single node indefinitely is at risk of:

Reviewer 2: complete sentence something like
```suggestion
A Two-Node with Fencing (TNF) cluster running on a single node indefinitely is at risk of experiencing the following issues:
```

* Complete cluster loss if the surviving node fails
* Inability to apply security updates or upgrades
* Certificate expiry if the cluster remains degraded beyond certificate validity periods
====

.Procedure

. Power on the failed node and verify network connectivity on all three network planes before the node can fully rejoin the cluster.
  - BMC or management network: The peer node must reach the BMC address of the failed node. Without BMC connectivity, fencing cannot protect the cluster if a subsequent failure occurs.
  - Cluster network: Bilateral connectivity between nodes for Corosync membership  (ports 5405 to 5407), Pacemaker management (port 2224), and etcd replication (ports 2379, 2380).

>   - Kubernetes API: The recovering node must be able to reach the API server (port 6443).

Reviewer 2: Use ** for unordered list

. Wait for Pacemaker to re-establish communication. Corosync detects the returning node and Pacemaker marks it as `ONLINE`. Pacemaker then starts managed resources, including kubelet and etcd, on the returning node.

. Wait for etcd to re-join the cluster. The returning member first rejoins as a learner (non-voting), receives a data snapshot, and replicates the log from the surviving member. Once caught up, it is promoted to a voting member, and the cluster returns to two-member quorum.
+
[NOTE]
====
This process typically takes about 15 to 25 minutes.
====

. Verify that the node transitions to `Ready` status:
+
[source,terminal]
----
$ oc get nodes
----

. Verify that cluster operators have cleared their degraded conditions:
+
[source,terminal]
----
$ oc get co

> ----

Reviewer 2: These would be part of verification

.Verification

After recovery, confirm the following:

* Both nodes show `Ready` status.
* The etcd operator no longer reports `Degraded=True`.
* Both fencing devices are operational.

> * Queued MachineConfig changes, upgrades, and certificate rotations can proceed.

Reviewer 2: Replace with specific steps/commands/actions

Reviewer: This command will capture the information for etcd Operator. The output is included. 

```
oc get co etcd 

output is 

NAME   VERSION       AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
etcd   4.22.0-rc.3   True        False         False      5d1h    
```

Reviewer: This is the check for the status of each node.

The command is `oc get nodes`

The output is
```
oc get nodes 
NAME                                                   STATUS   ROLES                         AGE    VERSION
e920t-01.cluster1.metal-platform.eng.rdu2.redhat.com   Ready    control-plane,master,worker   5d1h   v1.35.4
e920t-02.cluster1.metal-platform.eng.rdu2.redhat.com   Ready    control-plane,master,worker   5d1h   v1.35.4
```

Reviewer: An example of one of the nodes being in NotReady  Status

```
oc get nodes 
NAME                                                   STATUS     ROLES                         AGE    VERSION
e920t-01.cluster1.metal-platform.eng.rdu2.redhat.com   NotReady   control-plane,master,worker   5d1h   v1.35.4
e920t-02.cluster1.metal-platform.eng.rdu2.redhat.com   Ready      control-plane,master,worker   5d1h   v1.35.4
```

// Module included in the following assemblies:
// * installing_tnf/operating-degraded-two-node-fencing-cluster.adoc

:_mod-docs-content-type: PROCEDURE
[id="recovering-a-failed-tnf-node_{context}"]
= Recovering a failed TNF node

[role="_abstract"]
Recovery is automatic when the failed node is powered on and rejoins the network. 
Degraded operation is a temporary state. Restore the failed node as soon as possible to regain fault tolerance, fencing protection, and the ability to perform upgrades and maintenance. If the original node cannot be recovered, replace it.

[IMPORTANT]
====
A Two-Node with Fencing (TNF) cluster running on a single node indefinitely is at risk of experiencing the following issues:

* Complete cluster loss if the surviving node fails
* Inability to apply security updates or upgrades

> * Certificate expiry if the cluster remains degraded beyond certificate validity periods

Reviewer 3: ```suggestion
* Certificate expiration if the cluster remains degraded beyond certificate validity periods
```

====

.Procedure

. Power on the failed node and verify network connectivity on all three network planes before the node can fully rejoin the cluster.
  ** BMC or management network: The peer node must reach the BMC address of the failed node. Without BMC connectivity, fencing cannot protect the cluster if a subsequent failure occurs.
  ** Cluster network: Bilateral connectivity between nodes for Corosync membership  (ports 5405 to 5407), Pacemaker management (port 2224), and etcd replication (ports 2379, 2380).

>   ** Kubernetes API: The recovering node must be able to reach the API server (port 6443).

Reviewer 3: This seems to need more, especially the second bullet which isn't a full sentence. I am guessing here.

```suggestion
. Power on the failed node and verify network connectivity on all three network planes before the node can fully rejoin the cluster.
  ** BMC or management network: Ensure that the peer node must reach the BMC address of the failed node. Without BMC connectivity, fencing cannot protect the cluster if a subsequent failure occurs.
  ** Cluster network: Ensure there is bilateral connectivity between nodes for Corosync membership  (ports 5405 to 5407), Pacemaker management (port 2224), and etcd replication (ports 2379, 2380).
  ** Kubernetes API: Ensure that the recovering node is able to reach the API server (port 6443).
```

Author: ** Kubernetes API: Ensure that the recovering node can reach the API server (port 6443).

. Wait for Pacemaker to re-establish communication. Corosync detects the returning node and Pacemaker marks it as `ONLINE`. Pacemaker then starts managed resources, including kubelet and etcd, on the returning node.

> . Wait for etcd to re-join the cluster. The returning member first rejoins as a learner (non-voting), receives a data snapshot, and replicates the log from the surviving member. Once caught up, it is promoted to a voting member, and the cluster returns to two-member quorum.

Reviewer 3: ISG: [Do not use (once) as a conjunction to mean after or when.](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#o)

```suggestion
. Wait for etcd to re-join the cluster. The returning member first rejoins as a learner (non-voting), receives a data snapshot, and replicates the log from the surviving member. After it has caught up, it is promoted to a voting member, and the cluster returns to two-member quorum.
```

+
[NOTE]
====
This process typically takes about 15 to 25 minutes.
====

. Verify that the node transitions to `Ready` status:
+
[source,terminal]
----
$ oc get nodes
----

.Verification

. Verify that cluster Operators have cleared their degraded conditions:
+
[source,terminal]
----
$ oc get co
----
+

> . After recovery, confirm the following:

Reviewer 3: Instances of  _the following_ needs a noun. 
```suggestion
. After recovery, confirm the following conditions:
```

+

> * Both nodes show `Ready` status.

Reviewer 3: ```suggestion
* Verify that both nodes show `Ready` status by running the following command:
```

+
[source,terminal]
----
$ oc get nodes
----
+

> See the following sample output:

Reviewer 3: ```suggestion
The output is similar to the following example:
```

+
[source,terminal]
----
oc get nodes 
NAME                                                   STATUS   ROLES                         AGE    VERSION
e920t-01.cluster1.metal-platform.eng.rdu2.redhat.com   Ready    control-plane,master,worker   5d1h   v1.35.4
e920t-02.cluster1.metal-platform.eng.rdu2.redhat.com   Ready    control-plane,master,worker   5d1h   v1.35.4
----
+

> * The etcd Operator no longer reports `Degraded=True`.

Reviewer 3: ```suggestion
* Verify that the etcd Operator no longer reports `Degraded=True`  by running the following command:.
```

Author: As these conditions are under the same step, this suggestion is not applicable.

+
[source,terminal]
----
$ oc get co etcd 
----
+

> See the following sample output:

Reviewer 3: ```suggestion
The output is similar to the following example:
```

+
[source,terminal]
----
NAME   VERSION       AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
etcd   4.22.0-rc.3   True        False         False      5d1h
----
+  

> * Both fencing devices are operational.

Reviewer 3: ```suggestion
* Verify that both fencing devices are operational by running the following command:.
```
